source/adapters/opencl/kernel.cpp

```diff
@@ -250,13 +272,15 @@ urKernelGetSubGroupInfo(ur_kernel_handle_t hKernel, ur_device_handle_t hDevice,
 }
 
 UR_APIEXPORT ur_result_t UR_APICALL urKernelRetain(ur_kernel_handle_t hKernel) {
-  CL_RETURN_ON_FAILURE(clRetainKernel(cl_adapter::cast<cl_kernel>(hKernel)));
+  hKernel->incrementReferenceCount();
   return UR_RESULT_SUCCESS;
 }
 
 UR_APIEXPORT ur_result_t UR_APICALL
 urKernelRelease(ur_kernel_handle_t hKernel) {
-  CL_RETURN_ON_FAILURE(clReleaseKernel(cl_adapter::cast<cl_kernel>(hKernel)));
+  if (hKernel->decrementReferenceCount() == 0) {
```

Comment:
Shouldn't this code block be inside a mutex? I don't think using atomics is sufficient here. We can get into scenarios where 2 threads decrement the reference count at the same time (and before the check for == 0). That would cause a memory leak because the reference count will be -1.

Reply:
I think with this scenario one of them only will delete hKernel resources, and the other that will have a `-1` will already have a deleted resource, and both will return as they release successfully and with no memory leak. The atomic will only prevent them from checking at the same time, and prevent them from delete the resources both of them.